[UnifiedPDF] [iOS] Hook up accessibility for Unified PDF

ginalined opened the pull request:
#### a29783762efc4e2780f77187cd78a58ec3205c26
<pre>
[UnifiedPDF] [iOS] Hook up accessibility for Unified PDF
<a href="https://bugs.webkit.org/show_bug.cgi?id=288135">https://bugs.webkit.org/show_bug.cgi?id=288135</a>
<a href="https://rdar.apple.com/145235024">rdar://145235024</a>

Reviewed by Abrar Rahman Protyasha.

Add accessibility support for Unified PDF on iOS

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::convertFromPDFPageToScreenForAccessibility const):
(WebKit::UnifiedPDFPlugin::accessibilityObject const):
(WebKit::UnifiedPDFPlugin::accessibilityHitTestInPage):
* Source/WebKit/WebProcess/Plugins/PDF/WKAccessibilityPDFDocumentObject.h:
* Source/WebKit/WebProcess/Plugins/PDF/WKAccessibilityPDFDocumentObject.mm:
(-[WKAccessibilityPDFDocumentObject initWithPDFDocument:andElement:]):
(-[WKAccessibilityPDFDocumentObject setParent:]):
(-[WKAccessibilityPDFDocumentObject document]):
(-[WKAccessibilityPDFDocumentObject convertFromPDFPageToScreenForAccessibility:pageIndex:]):
(-[WKAccessibilityPDFDocumentObject setAXElements]):
(-[WKAccessibilityPDFDocumentObject accessibilityHitTest:]):
(-[WKAccessibilityPDFDocumentObject accessibilityElements]):
(-[WKAccessibilityPDFDocumentObject accessibilityScroll:]):
* Source/WebKit/WebProcess/WebPage/WKAccessibilityWebPageObjectIOS.mm:
(-[WKAccessibilityWebPageObject accessibilityHitTest:]):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::accessibilityObjectForMainFramePlugin):

Canonical link: <a href="https://commits.webkit.org/291103@main">https://commits.webkit.org/291103@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eebbd1ed0aa7b7383bf2718387e978c185efa959

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91941 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11468 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1016 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96876 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42546 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11819 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19957 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70550 "Found 1 new test failure: imported/w3c/web-platform-tests/mixed-content/gen/top.http-rp/opt-in/video-tag.https.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28036 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94942 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9010 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83274 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50878 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8740 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41761 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79062 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/869 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98905 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19063 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14160 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79575 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19315 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79120 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78801 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19520 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23344 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/650 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12101 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19044 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/24253 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18741 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22200 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20492 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -694,6 +694,12 @@ static inline FloatRect adjustExposedRectForNewScale(const FloatRect& exposedRec
+    if (!m_page)
+        return nil;
+
+    if (auto* pluginView = mainFramePlugIn())
```
```suggestion
    if (RefPtr pluginView = mainFramePlugIn())
```

pxlcoder commented on Source/WebKit/WebProcess/Plugins/PDF/WKAccessibilityPDFDocumentObject.mm:
```diff
@@ -70,6 +84,72 @@ - (void)setPDFDocument:(RetainPtr<PDFDocument>)document
+
+    auto pageCount = [_pdfDocument pageCount];
+    for (NSUInteger pageIndex = 0; pageIndex < pageCount; pageIndex ++) {
+        PDFPage *page = [_pdfDocument pageAtIndex:pageIndex];
```
```suggestion
        RetainPtr page = [_pdfDocument pageAtIndex:pageIndex];
```

pxlcoder commented on Source/WebKit/WebProcess/Plugins/PDF/WKAccessibilityPDFDocumentObject.h:
```diff
@@ -44,6 +44,7 @@ class WeakPtrImplWithEventTargetData;
     RetainPtr<PDFDocument> _pdfDocument;
     WeakObjCPtr<NSObject> _parent;
     ThreadSafeWeakPtr<WebKit::UnifiedPDFPlugin> _pdfPlugin;
+    NSMutableArray *_axElements;
```
```suggestion
    RetainPtr<NSMutableArray> _axElements;
```

pxlcoder commented on Source/WebKit/WebProcess/Plugins/PDF/WKAccessibilityPDFDocumentObject.mm:
```diff
@@ -35,28 +35,42 @@
 
     _pdfDocument = document;
     _pluginElement = element;
+    _axElements = [[NSMutableArray alloc] init];
```
```suggestion
    _axElements = adoptNS([[NSMutableArray alloc] init]);
```

pxlcoder commented on Source/WebKit/WebProcess/Plugins/PDF/WKAccessibilityPDFDocumentObject.mm:
```diff
@@ -70,6 +84,72 @@ - (void)setPDFDocument:(RetainPtr<PDFDocument>)document
+    if (_axElements.count)
+        [self setAXElements];
+
+    return _axElements;
```
After making `_axElements` a `RetainPtr`:
```suggestion
    if ([_axElements count])
        [self setAXElements];

    return _axElements.get();
```

twilco commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3588,9 +3591,22 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
+
+    auto [page, pointInPage] = rootViewToPage(pageRoot->chrome().screenToRootView(WebCore::IntPoint(point)));
+
+    if (!page)
```
Small style thing, but wondering if the new line separating `if (!page)` and `auto [page, pointInPage]` is necessary?

twilco commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3588,9 +3591,22 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
+
+    if ([page respondsToSelector:@selector(accessibilityHitTest:withPlugin:)])
+        return [page accessibilityHitTest:point withPlugin:m_accessibilityDocumentObject.get()];
     return nil;
```
Could write this a bit more succinctly like:

```cpp
auto [page, pointInPage] = rootViewToPage(pageRoot->chrome().screenToRootView(WebCore::IntPoint(point)));
if (page && [page respondsToSelector:@selector(accessibilityHitTest:withPlugin:)])
    return [page accessibilityHitTest:point withPlugin:m_accessibilityDocumentObject.get()];
return nil;
```

Also it's interesting that we get `pointInPage` but never use it, instead passing the original `point` to `accessibilityHitTest`. Is that correct?

ginalined replied:
Yes, because in pdfkit, we need to convert the rect of pdf elements from pdfPageView to screenView anyway to return the correct accessibilityFrame.

twilco commented on Source/WebKit/WebProcess/Plugins/PDF/WKAccessibilityPDFDocumentObject.mm:
```diff
@@ -70,6 +84,72 @@ - (void)setPDFDocument:(RetainPtr<PDFDocument>)document
+    return _pdfDocument.get();
+}
+
+- (NSRect)convertFromPDFPageToScreenForAccessibility:(NSRect)rectInPageCoordinate pageIndex:(WebKit::PDFDocumentLayout::PageIndex)pageIndex
```
I think "coordinates" is typically plural — i.e., `rectInPageCoordinates` rather than `rectInPageCoordinate`?

twilco commented on Source/WebKit/WebProcess/Plugins/PDF/WKAccessibilityPDFDocumentObject.mm:
```diff
@@ -70,6 +84,72 @@ - (void)setPDFDocument:(RetainPtr<PDFDocument>)document
+- (id)accessibilityHitTest:(NSPoint)point
+{
+
+    if (!_pdfDocument) {
```
I think the newline before the `if` may be unnecessary.

twilco commented on Source/WebKit/WebProcess/Plugins/PDF/WKAccessibilityPDFDocumentObject.mm:
```diff
@@ -70,6 +84,72 @@ - (void)setPDFDocument:(RetainPtr<PDFDocument>)document
+    }
+
+
+    if (RefPtr plugin = _pdfPlugin.get())
```
Extra newline here (there are two separating these blocks)

twilco commented on Source/WebKit/WebProcess/Plugins/PDF/WKAccessibilityPDFDocumentObject.mm:
```diff
@@ -297,6 +360,9 @@ - (void)gotoDestination:(PDFDestination *)destination
     });
 }
+
+#endif
```
Seems like we don't follow this convention elsewhere in this file, but I find it nice to include the compiler flag associated with this `#endif` in a comment like so:

```cpp
#endif // PLATFORM(FOO)
```

twilco commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3596,11 +3600,22 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
+        return nil;
+
+    auto [page, pointInPage] = rootViewToPage(pageRoot->chrome().screenToRootView(WebCore::IntPoint(point)));
+    if (page && [page respondsToSelector:@selector(accessibilityHitTest:withPlugin:)])
```
Do we need to nil-check `page`? Won't `respondsToSelector` return `NO` if `page` is nil?

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -181,12 +187,10 @@ - (void)formChanged:(NSNotification *)notification
-#if PLATFORM(MAC)
     m_accessibilityDocumentObject = adoptNS([[WKAccessibilityPDFDocumentObject alloc] initWithPDFDocument:m_pdfDocument andElement:&element]);
     [m_accessibilityDocumentObject setPDFPlugin:this];
     if (this->isFullFramePlugin() && m_frame && m_frame->page() && m_frame->isMainFrame())
```
Nit: I know you didn't write this line for this patch, but we do have a helper to simplify this line a bit
```suggestion
    if (isFullMainFramePlugin())
```

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3589,8 +3581,20 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
 
+FloatRect UnifiedPDFPlugin::convertFromPDFPageToScreenForAccessibility(const FloatRect& rectInPageCoordinates, PDFDocumentLayout::PageIndex pageIndex) const
+{
+    return WebCore::Accessibility::retrieveValueFromMainThread<FloatRect>([&]() ->FloatRect {
```
Nit: we use C++23, so no need for the empty arg list
```suggestion
    return WebCore::Accessibility::retrieveValueFromMainThread<FloatRect>([&] -> FloatRect {
```

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3602,11 +3606,22 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
+#if !PLATFORM(MAC)
+id UnifiedPDFPlugin::accessibilityHitTestInPage(WebCore::FloatPoint point)
+{
+    RefPtr pageRoot = this->page();
```
We usually prefix with `core` to disambiguate a WebCore object
```suggestion
    RefPtr corePage = this->page();
```

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3602,11 +3606,22 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
+}
+
+#if !PLATFORM(MAC)
+id UnifiedPDFPlugin::accessibilityHitTestInPage(WebCore::FloatPoint point)
```
I wish `accessibilityHitTestInPage` and `accessibilityHitTestIntPoint` were disambiguated a little better. Currently I can't tell why they're platform specific

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/WKAccessibilityPDFDocumentObject.mm:
```diff
@@ -70,6 +85,69 @@ - (void)setPDFDocument:(RetainPtr<PDFDocument>)document
+- (NSArray *)accessibilityElements
+{
+    if ([_axElements count])
+        [self setAXElements];
```
Did you mean to call `setAXElements` if `_axElements` is already non-empty? I thought it would read more like
```suggestion
    if (![_axElements count])
        [self setAXElements];
```

aprotyas commented on Source/WebKit/WebProcess/WebPage/WKAccessibilityWebPageObjectIOS.mm:
```diff
@@ -75,10 +75,16 @@ - (id)accessibilityHitTest:(NSPoint)point
 
+    RefPtr focusedLocalFrame = [self focusedLocalFrame];
+
+#if ENABLE(UNIFIED_PDF)
```
Better to go with `ENABLE(PDF_PLUGIN)`, since this is an iOS-only file anyway, and we plan to smoosh `UNIFIED_PDF` and `PDF_PLUGIN` together moving forward.

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/WKAccessibilityPDFDocumentObject.mm:
```diff
@@ -297,6 +358,9 @@ - (void)gotoDestination:(PDFDestination *)destination
 @end
 
-#endif
+#endif // UNIFIED_PDF
```
```suggestion
#endif // ENABLE(UNIFIED_PDF)
```

Commit:
[UnifiedPDF] [iOS] Hook up accessibility for Unified PDF
https://bugs.webkit.org/show_bug.cgi?id=288135
rdar://145235024

Reviewed by Abrar Rahman Protyasha.

Add accessibility support for Unified PDF on iOS

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::convertFromPDFPageToScreenForAccessibility const):
(WebKit::UnifiedPDFPlugin::accessibilityObject const):
(WebKit::UnifiedPDFPlugin::accessibilityHitTestInPage):
* Source/WebKit/WebProcess/Plugins/PDF/WKAccessibilityPDFDocumentObject.h:
* Source/WebKit/WebProcess/Plugins/PDF/WKAccessibilityPDFDocumentObject.mm:
(-[WKAccessibilityPDFDocumentObject initWithPDFDocument:andElement:]):
(-[WKAccessibilityPDFDocumentObject setParent:]):
(-[WKAccessibilityPDFDocumentObject document]):
(-[WKAccessibilityPDFDocumentObject convertFromPDFPageToScreenForAccessibility:pageIndex:]):
(-[WKAccessibilityPDFDocumentObject setAXElements]):
(-[WKAccessibilityPDFDocumentObject accessibilityHitTest:]):
(-[WKAccessibilityPDFDocumentObject accessibilityElements]):
(-[WKAccessibilityPDFDocumentObject accessibilityScroll:]):
* Source/WebKit/WebProcess/WebPage/WKAccessibilityWebPageObjectIOS.mm:
(-[WKAccessibilityWebPageObject accessibilityHitTest:]):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::accessibilityObjectForMainFramePlugin):

Canonical link: https://commits.webkit.org/291103@main
